src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -1894,5 +1882,232 @@ struct S
+        {
+            case [42]:
+            case [..,42]:
+                break;
+        }
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRange(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void Subsumption_07()
+        {
+            var src = @"
+class C
+{
+    void Test(int[] a)
+    {
+        switch (a)
+        {
+            case [>0, ..]:
+            case [.., <=0]:
+            case [var unreachable]:
+                    break;
+        }
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRange(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics(
+                // (10,18): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match.
+                //             case [var unreachable]:
+                Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, "[var unreachable]").WithLocation(10, 18));
+        }
+
+        [Theory]
+        [CombinatorialData]
+        public void Subsumption_Slice_00(
+            [CombinatorialValues(
+                "[1,2,3]",
+                "[1,2,3,..[]]",
+                "[1,2,..[],3]",
+                "[1,..[],2,3]",
+                "[..[],1,2,3]",
+                "[1,..[2,3]]",
+                "[..[1,2],3]",
+                "[..[1,2,3]]",
+                "[..[..[1,2,3]]]",
+                "[..[1,2,3,..[]]]",
+                "[..[1,2,..[],3]]",
+                "[..[1,..[],2,3]]",
+                "[..[..[],1,2,3]]",
+                "[..[1,..[2,3]]]",
+                "[..[..[1,2],3]]")]
+            string case1,
+            [CombinatorialValues(
+                "[1,2,3]",
+                "[1,2,3,..[]]",
+                "[1,2,..[],3]",
+                "[1,..[],2,3]",
+                "[..[],1,2,3]",
+                "[1,..[2,3]]",
+                "[..[1,2],3]",
+                "[..[1,2,3]]",
+                "[..[..[1,2,3]]]",
+                "[..[1,2,3,..[]]]",
+                "[..[1,2,..[],3]]",
+                "[..[1,..[],2,3]]",
+                "[..[..[],1,2,3]]",
+                "[..[1,..[2,3]]]",
+                "[..[..[1,2],3]]")]
+            string case2)
+        {
+            var src = @"
+using System;
+class C
+{
+    void Test(Span<int> a)
+    {
+        switch (a)
+        {
+            case " + case1 + @":
+            case " + case2 + @":
+                break;
+        }
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRangeAndSpan(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics(
+                // (10,18): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match.
+                Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, case2).WithLocation(10, 18)
+                );
+        }
```

Comment:
We may need to update the spec to explain expectations on length.

Just brainstorming before having looked at the spec or code change, consider some more tests for subsumption on length:
- `case [..]: case { Length: -1 }:` // direct check on length, does the "length is positive" assumption hold?
- `case { Length: >= 0 }:`
- `case []: case [var one]: case { Length: > 1 }` // mixing direct and indirect length checks

Subsumption and exhaustiveness (for switch expressions) are closely related. Is your plan to cover the switch expression exhaustiveness in this PR later on?

For subsumption on nested slices, consider testing examples with designations: `case [var one]: case [..[var alsoOne]]:`.

Reply:
>exhaustiveness

that would be the result of subsumption checking, so we definitely need to be cover it at the same time.

Reply:
Another test idea: `case [string, ..]: case [.., not string]: case [var x]:`

Reply:
>direct check on length

@jcouv This part is changed earlier (extended property patterns) so we need to update the branch. 

Would you prefer to do the merge yourself?

Reply:
I'll go ahead and refresh the feature branch.

Reply:
Another test idea (works fine although I still need to figure out how ;-) ): `case [_, > 0, ..]: case [.., <= 0, _]: case [_, var unreachable, _]:`

Reply:
In inspecting the scenario `case [_, > 0, ..]: case [.., <= 0, _]: case [_, var unreachable, _]:` I noticed something strange. The second case ends up with a `length == 3` test, which seems wrong. See comment below for the state where this happens.

I've confirmed there is a problem by verifying the execution of the following test. 
This makes me think we should add code execution to those subsumption tests as well.

```
        [Fact]
        public void Subsumption_TODO2()
        {
            var src = @"
C.Test(new[] { 42, -1, 0, 42 });

class C
{
    public static void Test(int[] a)
    {
        switch (a)
        {
            case [_, > 0, ..]: 
                System.Console.Write(1);
                break;
            case [.., <= 0, _]: 
                System.Console.Write(2);
                break;
            default:
                System.Console.Write(3);
                break;
        }
    }
}";
            var comp = CreateCompilationWithIndexAndRange(src, parseOptions: TestOptions.RegularWithListPatterns);
            comp.VerifyEmitDiagnostics();
            CompileAndVerify(comp, expectedOutput: "2"); // incorrectly outputs 3
        }
```

```
State 0
    1. [case [_, > 0, ..]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    3. [case [_, var unreachable, _]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagValueTest(t1 == ConstantValueI32(3: Int32)), t2=DagIndexerEvaluation(t0[1])) BIND[unreachable=t2]
    Test: ?DagNonNullTest(t0)
    TrueBranch: 1
    FalseBranch: 9
State 1
    1. [case [_, > 0, ..]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    3. [case [_, var unreachable, _]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagValueTest(t1 == ConstantValueI32(3: Int32)), t2=DagIndexerEvaluation(t0[1])) BIND[unreachable=t2]
    Test: t1=DagPropertyEvaluation(t0.Length)
    TrueBranch: 2
State 2
    1. [case [_, > 0, ..]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    3. [case [_, var unreachable, _]:] AND(?DagValueTest(t1 == ConstantValueI32(3: Int32)), t2=DagIndexerEvaluation(t0[1])) BIND[unreachable=t2]
    Test: ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32))
    TrueBranch: 3
    FalseBranch: 9
State 3 REMAINING t1:[2..2147483647]
    1. [case [_, > 0, ..]:] AND(t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    3. [case [_, var unreachable, _]:] AND(?DagValueTest(t1 == ConstantValueI32(3: Int32)), t2=DagIndexerEvaluation(t0[1])) BIND[unreachable=t2]
    Test: t2=DagIndexerEvaluation(t0[1])
    TrueBranch: 4
State 4 REMAINING t1:[2..2147483647]
    1. [case [_, > 0, ..]:] ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32))
    2. [case [.., <= 0, _]:] AND(t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    3. [case [_, var unreachable, _]:] ?DagValueTest(t1 == ConstantValueI32(3: Int32)) BIND[unreachable=t2]
    Test: ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32))
    TrueBranch: 5
    FalseBranch: 6
*State 5 REMAINING t1:[2..2147483647] t2:[1..2147483647]
    1. [case [_, > 0, ..]:] MATCH
State 6 REMAINING t1:[2..2147483647] t2:[-2147483648..0]
    2. [case [.., <= 0, _]:] AND(t3=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t1 == ConstantValueI32(3: Int32))) // I don't understand why we have a test for `t1 == 3` (meaning lenght is 3)
    3. [case [_, var unreachable, _]:] ?DagValueTest(t1 == ConstantValueI32(3: Int32)) BIND[unreachable=t2]
    Test: t3=DagIndexerEvaluation(t0[-2])
    TrueBranch: 7
State 7 REMAINING t1:[2..2147483647] t2:[-2147483648..0]
    2. [case [.., <= 0, _]:] ?DagValueTest(t1 == ConstantValueI32(3: Int32))
    3. [case [_, var unreachable, _]:] ?DagValueTest(t1 == ConstantValueI32(3: Int32)) BIND[unreachable=t2]
    Test: ?DagValueTest(t1 == ConstantValueI32(3: Int32))
    TrueBranch: 8
    FalseBranch: 9
*State 8 REMAINING t1:[3..3] t2:[-2147483648..0]
    2. [case [.., <= 0, _]:] MATCH
*State 9 FAIL
```

Reply:
<details><summary>new decision dag dump</summary>

```
State 0
    1. [case [_, > 0, ..]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: ?DagNonNullTest(t0)
    TrueBranch: 1
    FalseBranch: 8
State 1
    1. [case [_, > 0, ..]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: t1=DagPropertyEvaluation(t0.Length)
    TrueBranch: 2
State 2
    1. [case [_, > 0, ..]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(2: Int32)), t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: ?DagRelationalTest(t1 >= ConstantValueI32(2: Int32))
    TrueBranch: 3
    FalseBranch: 8
State 3 REMAINING t1:[2..2147483647]
    1. [case [_, > 0, ..]:] AND(t2=DagIndexerEvaluation(t0[1]), ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32)))
    2. [case [.., <= 0, _]:] AND(t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: t2=DagIndexerEvaluation(t0[1])
    TrueBranch: 4
State 4 REMAINING t1:[2..2147483647]
    1. [case [_, > 0, ..]:] ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32))
    2. [case [.., <= 0, _]:] AND(t3=DagIndexerEvaluation(t0[-2]), ?DagRelationalTest(t3 <= ConstantValueDefault(0: Int32)))
    Test: ?DagRelationalTest(t2 > ConstantValueDefault(0: Int32))
    TrueBranch: 5
    FalseBranch: 6
*State 5 REMAINING t2:[1..2147483647] t1:[2..2147483647]
    1. [case [_, > 0, ..]:] MATCH
State 6 REMAINING t2:[-2147483648..0] t1:[2..2147483647]
    2. [case [.., <= 0, _]:] t3=DagIndexerEvaluation(t0[-2])
    Test: t3=DagIndexerEvaluation(t0[-2])
    TrueBranch: 7
*State 7 REMAINING t2:[-2147483648..0] t1:[2..2147483647]
    2. [case [.., <= 0, _]:] MATCH
*State 8 FAIL
```
</details>

Reply:
Another test idea: comparing the exhaustiveness behavior of `case { Length: 0 }: case { Length: 1 }: case { Length: >1 }:` compared to `case []: case [_]: [_, ..]:`.
My understanding is that they'll behave differently, but that still seems strange to me. We should probably discuss in the context of spec'ing this out.

Reply:
They would behave exactly the same. Added a test to verify.